doc/man/hawk.1
'\" t
.\"     Title: hawk
.\"    Author: [see the "Authors" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 02/12/2020
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "HAWK" "1" "02/12/2020" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
hawk \- the image reconstruction package
.SH "SYNOPSIS"
.sp
.nf
\fIhawk\fR [\-\-version] [\-\-help] <command> [<args>]
.fi
.SH "DESCRIPTION"
.sp
Hawk is a image reconstruction package that contains the tools necessary to process and phase a continuous diffraction pattern\&. It is typically used with diffraction patterns obtained from X\-ray single particle diffraction experiments, such as those performed at X\-ray free\-electron lasers and table\-top X\-ray lasers, but the algorithms should work for other diffraction experiments such as electron or optical diffraction\&.
.SH "OPTIONS"
.PP
\fB\-\-version\fR
.RS 4
Prints the Hawk package version\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Prints the synopsis along with a list and short explanation of the possible commands\&. When given a <command> it will show the man page of that <command>, for example
\fBhawk \-\-help convert\fR
will show the man page for hawk\-convert\&.
.RE
.SH "COMMANDS"
.PP
\fBconvert\fR
.RS 4
Converts files between different file formats\&.
.RE
.PP
\fBphase\fR
.RS 4
Phase a diffraction pattern\&.
.RE
.PP
\fBprepare\fR
.RS 4
Prepare an image for phasing\&.
.RE
.SH "AUTHORS"
.sp
Hawk was created by Filipe Maia and includes contributions from Tomas Ekeberg, Max Hantke and Gijs van der Schot\&.
.SH "REPORTING BUGS"
.sp
Please report bugs on https://github\&.com/FilipeMaia/hawk/issues\&.
.SH "SEE ALSO"
.sp
\fBhawk\-convert\fR(1) \fBhawk\-phase\fR(1) \fBhawk\-prepare\fR(1)


Changelog
2009-04-23	Filipe Maia	<filipe@xray.bmc.uu.se>
	* Made everything compile in windows
	* New alpha quality graphical interface
	* New gtm like dummy package
	* Made self contained packages for windows linux and macosx
2008-08-16	Filipe Maia	<filipe@xray.bmc.uu.se>
	* Added the capability to work with 3D images.
	* Added many new options to control the support update mechanism.
	* Added a metadata options file which includes some options documentation
	* Added more examples and a few preliminary benchmarks
	* Added the option of comparing the current solution with an expected solution using
	the solution_file option.
	* Added the possibility to filter the intensities at runtime according to the support
	using the filter_intensities option.
	* Changed the way images are rescaled so that now the image guess is scaled to the
	experimental intensities so that they both have the same power (sum of squares)
	* Added the option to specify a file with the standard deviations of each pixel
	for the intensity file, using the intensities_std_dev_file option.
	* Added the option to specify the support of the autocorrelation for denoising
	purposes using the autocorrelation_support_file option.
	* Added a two graphical programs to look at images. Look and Preview. They can
	also be used for some basic manipulation.
	* Added a program to convert images to reflection lists.
	* Many other small improvements and code cleaning.
2007-04-19	Filipe Maia	<filipe@xray.bmc.uu.se>

	* Added the image_blur_radius and image_blur_period options.
	When image_blur_period is different than 0, the real image
	image will be gaussian blurred using a standard deviation
	of image_blur_radius every image_blur_period iteration.

2007-04-19	Filipe Maia	<filipe@xray.bmc.uu.se>

	* Introduced the "object_area" option, which is a floating point
	number which quantifies what's the area of the object compared
	to the size of the image. It can take values between 0.0 and 1.0.
	* Added a new support update algorithm called "constant_area".
	It simply keeps the area of the support fixed. The support area
	is given by the new "object_area" option.

2007-04-18	Filipe Maia	<filipe@xray.bmc.uu.se>

	* Fixed the fourier space error calculation for the RAAR algorithm.

2007-04-13	Filipe Maia	<filipe@xray.bmc.uu.se>

	* Changed the boolean "blur_radius" option to a float "patterson_blur_radius"
	that allows the set also the amount of blurring.
	* Fixed a bug in the initial estimation of the support, caused by incorrect
	calculation of the autocorrelation.
	* Split the support functions and algorithm functions in 2 new files.

2007-04-12	Filipe Maia	<filipe@xray.bmc.uu.se>

	* Added a max_iterations option to uwrapc.conf that makes the program stop
	after that number of iterations has been reached.
	* Changed the -C option in process_image. Now it average the entire image
	assuming centrosymmetry.

2007-03-28	Filipe Maia	<filipe@xray.bmc.uu.se>

	* Public svn repository at svn://130.238.41.203/hawk

2007-03-27	Filipe Maia	<filipe@xray.bmc.uu.se>

	* Fixed a critical bug in tiff2h5. The width and height were
	not being set before the autocorrelation was calculated.

2007-03-21	Filipe Maia	<filipe@xray.bmc.uu.se>

	* Begining of changelog
	* Added a new utility called denoise that tries to remove some poisson
	noise from a picture. Please check utils/denois.c for details.
	* Added background_subtract which tries to subtract the background of a
	series of similar pictures. It's too specific for general usage though.
	* Started to build static binaries for easier installation.


src/hawk.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>

int convert_main(int argc, char ** argv);

char * version(void){
	static char buffer[9];
	int y = HAWK_VERSION/10000;
	int m = (HAWK_VERSION/100)%100;
	int d = HAWK_VERSION%100;
	sprintf(buffer,"%d.%d.%d",y,m,d);
	return buffer;
}



int main(int argc, char ** argv){
	static char help[] = 
	"usage: hawk [--version|--help] <command> [<args>]\n\
\n\
The most commonly used hawk commands are:\n\
   convert    Convert an image from a format to another\n\
   phase      Phase a diffraction pattern\n\
\n\
See 'hawk help <command>' for more information on a specific command.\n\
";

	if(argc == 1 || (argc == 2 && (strcmp(argv[1],"--help") == 0 || strcmp(argv[1],"help") == 0))){
		printf("%s\n",help);
		exit(0);
	}
	if(strcmp(argv[1],"--version") == 0|| strcmp(argv[1],"version") == 0){
		printf("hawk version %s\n",version());
		exit(0);
	}
	if(strcmp(argv[1],"help") == 0){		
		char buffer[1024];
		snprintf(buffer,1024,"hawk-%s",argv[2]);
		execlp("man", "man", buffer, (char *)NULL);
		/* If everything goes well execlp will not return */
		printf("failed to exec man: %s\n", strerror(errno));
		printf("Could not start \"man\"\n");
		exit(-1);
	}else if(strcmp(argv[1],"convert") == 0){
		convert_main(argc-1,&argv[1]);
	}else{
		printf("Unknown option: %s\n",argv[1]);
		printf("%s\n",help);
		exit(0);
	}
	return 0;
}